text/0113-secure-script-container.md

```diff
@@ -0,0 +1,183 @@
+
+A syntax like the following could be provided:
+
+```html
+<secure-script extra-globals="ga"></secure-script>
+```
+
+### Known Limitations
+
+* Currently there is an outstanding issue with synthetic shadow in that the Event object patching happens lazily, so we are unable to prevent its patching at the time of creating the secure container, which means that events are still retargeted. [See LWC PR #1569](https://github.com/salesforce/lwc/pull/1569).
+
+
+### Testing
+
+We will want to gather the top 20 common libraries and implement their behavior inside of `secure-script` to ensure that we are prooperly accounting for all these libraries' use cases. This will be our integration test suite. 
+
+### Prototype Code
+
+```js
+import createSecureEnvironment from '@caridy/sjs/lib/browser-realm';
+
+// creating the new sandbox with all global endowments
+const evaluate = createSecureEnvironment(
+    undefined,
+    window,
+);
+
+evaluate(`
+    // This initialization will prevent any of these APIs to be polyfilled
+    // on the blue realm that can affect this sandbox.
+    const pE = Object.preventExtensions;
+    pE(HTMLElement.prototype);
+    pE(Element.prototype);
+    pE(Node.prototype);
+    pE(Event.prototype);
+    pE(Document.prototype);
+    pE(EventTarget.prototype);
+    pE(MutationObserver.prototype);
+    pE(HTMLCollection.prototype);
+    pE(NodeList.prototype);
+    pE(ShadowRoot.prototype);
+    pE(HTMLSlotElement.prototype);
+    pE(Text.prototype);
+`);
+
+class SecureScript extends HTMLElement {
+    connectedCallback() {
+        const src = this.textContent;
+        if (src) {
+            evaluate(src);
+        }
+    }
+}
+
+customElements.define('secure-script', SecureScript);
+```
+
+## Drawbacks
+
+- We are potentially prolonging the life of our synthetic shadow dom polyfill, instead of encouraging community library authors to adopt shadow DOM semantics.
+- The element may end up being anywhere from 5kb - 20kb.
+
+## Alternatives
+
+- **Allow shadow DOM to be toggled at a more granular level.** There is a forthcoming RFC for allowing component authors to decide if they want to opt in to shadow DOM or not. This would mean we can keep top-level elements and layout components in the light DOM, but base components or managed components can choose to safeguard their internals. Until we allow this, `secure-script` serves as a mitigation for application developers looking to add third party libraries to their apps.
+- Do not apply the shadow DOM polyfill at all: this would allow component authors to violate shadow DOM encapsulation which is not an option.
+- If we don't do this, then companies / websites building on our platform will not be able to integrate with common third party libraries.
+
+## Adoption strategy
+
+TBD
```

Comment:
Is the expectation that the admin will click an application/extension for their site to add Google Analytics (for example) and it will be added by us into the `secure-script`? Or will the admins/devs place `secure-script` themselves and it may happen to be Google Analytics, Evergage,etc.